keras_nlp/metrics/edit_distance.py

```diff
@@ -0,0 +1,223 @@
+    ... )
+    >>> y = tf.strings.split(["the cat was found under the bed"])
+    >>> y = tf.strings.split(["the cat was found under the bed"])
+    >>> output = model.evaluate(y, x, return_dict=True)
+    >>> output["edit_distance"]
+    0.3636363744735718
+    """
+
+    def __init__(
+        self,
+        normalize=True,
+        dtype=None,
+        name="edit_distance",
+        **kwargs,
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+
+        if not tf.as_dtype(self.dtype).is_floating:
+            raise ValueError(
+                "`dtype` must be a floating point type. "
+                f"Received: dtype={dtype}"
+            )
+
+        self.normalize = normalize
+
+        self._aggregate_unnormalized_edit_distance = self.add_weight(
+            name="aggregate_unnormalized_edit_distance",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        if normalize:
+            self._aggregate_reference_length = self.add_weight(
+                name="aggregate_reference_length",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+        else:
+            self._number_of_samples = self.add_weight(
+                name="number_of_samples",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        def validate_and_fix_rank(inputs, tensor_name):
+            if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+                inputs = tf.ragged.constant(inputs)
+
+            if inputs.shape.rank == 1:
+                return tf.RaggedTensor.from_tensor(inputs[tf.newaxis])
+            elif inputs.shape.rank == 2:
+                return inputs
+            else:
+                raise ValueError(
+                    f"{tensor_name} must be of rank 1 or 2. "
+                    f"Found rank: {inputs.shape.rank}"
+                )
+
+        y_true = validate_and_fix_rank(y_true, "y_true")
+        y_pred = validate_and_fix_rank(y_pred, "y_pred")
+
+        if self.normalize:
+            self._aggregate_reference_length.assign_add(
+                tf.cast(tf.size(y_true.flat_values), dtype=self.dtype)
+            )
+
+        def calculate_edit_distance(args):
+            reference, hypothesis = args
+
+            reference = tf.sparse.from_dense([reference])
+            hypothesis = tf.sparse.from_dense([hypothesis])
+
+            edit_distance = tf.squeeze(
+                tf.edit_distance(
+                    hypothesis=hypothesis,
+                    truth=reference,
+                    normalize=False,
+                )
+            )
+
+            self._aggregate_unnormalized_edit_distance.assign_add(
+                tf.cast(edit_distance, dtype=self.dtype)
+            )
+            if not self.normalize:
+                self._number_of_samples.assign_add(tf.cast(1, dtype=self.dtype))
+            return 0
```

Comment:
Curious - is the return value required for `map_fn`? seems this return value is not being used

Reply:
The issue is that we can't return None. This is the function spec for `tf.map_fn`:

```
tf.map_fn(
    fn,
    elems,
    dtype=None,
    parallel_iterations=None,
    back_prop=True,
    swap_memory=False,
    infer_shape=True,
    name=None,
    fn_output_signature=None
)
```

In particular, have a look at `fn_output_signature`:
```
The output signature of fn. Must be specified if fn's input and output signatures are different (i.e., if their structures, dtypes, or tensor types do not match). fn_output_signature can be specified using any of the following:
A tf.DType or tf.TensorSpec (to describe a tf.Tensor)
A tf.RaggedTensorSpec (to describe a tf.RaggedTensor)
A tf.SparseTensorSpec (to describe a tf.sparse.SparseTensor)
A (possibly nested) tuple, list, or dict containing the above types.
```